hbase-rsgroup/src/test/java/org/apache/hadoop/hbase/rsgroup/TestRSGroupsAdmin2.java

```diff
@@ -459,4 +466,206 @@ public boolean evaluate() throws Exception {
+        LOG.debug("server table region size is:{}", regions.size());
+        assert regions.size() >= 1;
+        // when there is exactly one region left, we can determine the move operation encountered
+        // exception caused by the strange region state.
+        if (regions.size() == 1) {
+          assertEquals(regions.get(0).getRegionNameAsString(),
+              rsn.getRegionInfo().getRegionNameAsString());
+          rsn.setState(RegionState.State.OPEN);
+          LOG.info("set region {} state OPEN", rsn.getRegionInfo().getRegionNameAsString());
+          changed.set(true);
+          break;
+        }
+        sleep(5000);
+      }
+    });
+  }
+
+  @Test
+  public void testFailedMoveWhenMoveServer() throws Exception{
+    String groupName = getGroupName(name.getMethodName());
+    rsGroupAdmin.addRSGroup(groupName);
+    final RSGroupInfo newGroup = rsGroupAdmin.getRSGroupInfo(groupName);
+    Pair<ServerName, RegionStateNode> gotPair = createTableAndSetARegionState(newGroup,
+        10);
+    try{
+      rsGroupAdmin.moveServers(Sets.newHashSet(gotPair.getFirst().getAddress()),
+          newGroup.getName());
+      fail("move servers to group should fail");
+    }catch (IOException e){
+      assertTrue(e.getMessage().contains(
+          gotPair.getSecond().getRegionInfo().getRegionNameAsString()));
+    }
+  }
+
+  @Test
+  public void testFailedMoveWhenMoveTable() throws Exception{
+    final RSGroupInfo newGroup = addGroup(getGroupName(name.getMethodName()), 1);
+    Pair<ServerName, RegionStateNode> gotPair = createTableAndSetARegionState(newGroup, 5);
+    try{
+      rsGroupAdmin.moveTables(Sets.newHashSet(tableName), newGroup.getName());
+      fail("move tables to group should fail");
+    }catch (IOException e){
+      assertTrue(e.getMessage().contains(
+          gotPair.getSecond().getRegionInfo().getRegionNameAsString()));
+    }
+  }
+
+  private Pair<ServerName, RegionStateNode> createTableAndSetARegionState(RSGroupInfo rsGroupInfo,
+      int tableRegionCount) throws Exception{
+    final byte[] familyNameBytes = Bytes.toBytes("f");
+    // All the regions created below will be assigned to the default group.
+    TEST_UTIL.createMultiRegionTable(tableName, familyNameBytes, tableRegionCount);
+    TEST_UTIL.waitFor(WAIT_TIMEOUT, new Waiter.Predicate<Exception>() {
+      @Override
+      public boolean evaluate() throws Exception {
+        List<String> regions = getTableRegionMap().get(tableName);
+        if (regions == null) {
+          return false;
+        }
+        return getTableRegionMap().get(tableName).size() >= tableRegionCount;
+      }
+    });
+
+    return setARegionState(rsGroupInfo);
+  }
+
+  /**
+   * Randomly choose a region to set state.
+   * @param newGroup target group
+   * @return source server of region, and region state
+   * @throws IOException if methods called throw
+   */
+  private Pair<ServerName, RegionStateNode> setARegionState(RSGroupInfo newGroup)
```

Comment:
Small nit: maybe worth mention the state in the method name? something like _randomSetOneRegionToSplittingState_?

Reply:
Thanks, I'll change the method name.